File: elixir/meetup/meetup.exs
defmodule Meetup do
  @moduledoc """
  Calculate meetup dates.
  """

  @type weekday ::
      :monday | :tuesday | :wednesday
    | :thursday | :friday | :saturday | :sunday

  @type schedule :: :first | :second | :third | :fourth | :last | :teenth

  @doc """
  Calculate a meetup date.

  The schedule is in which week (1..4, last or "teenth") the meetup date should
  fall.
  """

  @spec meetup(pos_integer, pos_integer, weekday, schedule) :: :calendar.date
  def meetup(year, month, weekday, :teenth) do
    meetup(year, month, weekday, 13..19)
  end

  def meetup(year, month, weekday, :first) do
    meetup(year, month, weekday, 1..7)
  end

  def meetup(year, month, weekday, :second) do
    meetup(year, month, weekday, 8..14)
  end
  
  def meetup(year, month, weekday, :third) do
    meetup(year, month, weekday, 15..21)
  end

  def meetup(year, month, weekday, :fourth) do
    meetup(year, month, weekday, 22..28)
  end

  def meetup(year, month, weekday, :last) do
    last_day = Date.new(year, month, 1) |> elem(1) |> Date.days_in_month
    meetup(year, month, weekday, (last_day - 6)..last_day)
  end
    
  def meetup(year, month, weekday, range) do
    range
    |> Enum.to_list
    |> Enum.map(&(Date.new(year, month, &1)))
    |> Enum.find(fn({:ok, date}) -> valid?(date, weekday) end)
    |> elem(1)
    |> Date.to_erl
  end

  def valid?(date, :monday),    do: Date.day_of_week(date) === 1
  def valid?(date, :tuesday),   do: Date.day_of_week(date) === 2
  def valid?(date, :wednesday), do: Date.day_of_week(date) === 3
  def valid?(date, :thursday),  do: Date.day_of_week(date) === 4
  def valid?(date, :friday),    do: Date.day_of_week(date) === 5
  def valid?(date, :saturday),  do: Date.day_of_week(date) === 6
  def valid?(date, :sunday),    do: Date.day_of_week(date) === 7
end

File: python/meetup/meetup.py
import datetime
import calendar

def meetup_day(year, month, day_of_week, selection):

  if selection == '1st':
    day_range = range(1, 8)
  elif selection == '2nd':
    day_range = range(8, 15)
  elif selection == 'teenth':
    day_range = range(13, 20)
  elif selection == '3rd':
    day_range = range(15, 22)
  elif selection == '4th':
    day_range = range(22, 30)
  elif selection == 'last':
    if month == 2:
      if calendar.isleap(year):
        day_range = range(23, 30)
      else:
        day_range = range(22, 29)
    elif month == 3 | 4 | 6 | 9 :
      day_range = range(24, 31)
    else:
      day_range = range(25, 32)
  elif selection == '5th':
    day_range = range(30, 31)
  else:
    raise Exception
  

  for day in day_range:
    date = datetime.date(year, month, day)

    if day_of_week == "Monday" and date.weekday() == 0:
      return date
    elif day_of_week == "Tuesday" and date.weekday() == 1:
      return date
    elif day_of_week == "Wednesday" and date.weekday() == 2:
      return date
    elif day_of_week == "Thursday" and date.weekday() == 3:
      return date
    elif day_of_week == "Friday" and date.weekday() == 4:
      return date
    elif day_of_week == "Saturday" and date.weekday() == 5:
      return date
    elif day_of_week == "Sunday" and date.weekday() == 6:
      return date

  raise Exception
  

